gpu_direct_rdma_access.c

```diff
@@ -838,6 +855,26 @@ struct rdma_device *rdma_open_device_server(struct sockaddr *addr)
     return NULL;
 }
 
+int rdma_reset_device(struct rdma_device *device)
+{
+    struct ibv_qp_attr      qp_attr;
+    enum ibv_qp_attr_mask   attr_mask;
+    memset(&qp_attr, 0, sizeof qp_attr);
+    /* - - - - - - - Modify QP to RESET - - - - - - - */
+    qp_attr.qp_state = IBV_QPS_RESET;
+    attr_mask = IBV_QP_STATE;
+    DEBUG_LOG("ibv_modify_qp(qp = %p, qp_attr.qp_state = %d, attr_mask = 0x%x)\n",
+                    device->qp, qp_attr.qp_state, attr_mask);
+    if (ibv_modify_qp(device->qp, &qp_attr, attr_mask)) {
+        fprintf(stderr, "Failed to modify QP to RESET\n");
+        return 1;
+    }
+    DEBUG_LOG ("ibv_modify_qp to state %d completed.\n", qp_attr.qp_state, device->qp->qp_num);
+    
+    /* - - - - - - - Modify QP to RTS (RESET->INIT->RTR->RTS) - - - - - - - */
+    return setup_server_qp_state(device);
```

Comment:
what about client side calling this API?
at least verify this is a server rdma_device... don't assume it.
probably need to think hard when/what can cause the client side to get into error in the first place

Reply:
this condition was added
`(device->srq != NULL || !device->qpex || !device->mqpex) `

Reply:
I suggest to create an internal `static int is_server(rdma_device *device)`

device->srq != NULL is probably enough. no need to check ALL server conditions.